include/CObjectCache.h

```diff
@@ -70,191 +122,433 @@ namespace impl
+    struct CPreInsertionVerifier<ContainerT_T, ContainerT, true, IsAssocContainer>
+    {
+        template<typename ...Ts>
+        static bool verify(Ts...) { return true; }
+    };
+    template<template<typename...> class ContainerT_T, typename ContainerT>
+    struct CPreInsertionVerifier<ContainerT_T, ContainerT, false, false>
+    {
+        static bool verify(const ContainerT& _container, const typename ContainerT::iterator& _itr, const typename ContainerT::value_type::first_type& _key)
+        {
+            if (_itr != std::cend(_container) && !(_key < _itr->first)) // used `<` instead of `==` operator here to keep consistency with std::map (so key type doesn't need to define operator==)
+                return false;
 
-template<
-    typename K,
-    typename T,
-    template<typename...> class ContainerT_T
->
-class CObjectCache<K, T, ContainerT_T, true> : public impl::CObjectCacheBase<ContainerT_T, std::pair<K, T*>>
-{
-    using ValueType = std::pair<K, T*>;
+            return true;
+        }
+    };
+    template<template<typename...> class ContainerT_T, typename ContainerT>
+    struct CPreInsertionVerifier<ContainerT_T, ContainerT, false, true>
+    {
+        static bool verify(const std::pair<typename ContainerT::iterator, bool>& _insertionRes)
+        {
+            return _insertionRes.second;
+        }
+    };
 
-public:
-    inline explicit CObjectCache(const std::function<void(T*)>& _greeting, const std::function<void(T*)>& _disposal) : impl::CObjectCacheBase<ContainerT_T, std::pair<K, T*>>(_greeting, _disposal) {}
-    inline explicit CObjectCache(std::function<void(T*)>&& _greeting = nullptr, std::function<void(T*)>&& _disposal = nullptr) : impl::CObjectCacheBase<ContainerT_T, std::pair<K, T*>>(std::move(_greeting), std::move(_disposal)) {}
+    //! Use in non-static member functions
+#define MY_TYPE typename std::remove_reference<decltype(*this)>::type
+#define INSERT_IMPL_VEC \
+    const typename Base::PairType_impl newVal{ _key, _val };\
+    auto it = std::lower_bound(std::begin(this->m_container), std::end(this->m_container), newVal, [](const typename Base::PairType_impl& _a, const typename Base::PairType_impl& _b) -> bool {return _a.first < _b.first; });\
+    if (\
+    !impl::CPreInsertionVerifier<ContainerT_T, typename Base::ContainerT, std::is_base_of<impl::CMultiCache_tag, MY_TYPE>::value>::verify(this->m_container, it, _key)\
+    )\
+        return false;\
+    \
+    greet(newVal.second);\
+    this->m_container.insert(it, newVal);\
+    return true;
+#define INSERT_IMPL_ASSOC \
+    auto res = this->m_container.insert({ _key, _val });\
+    const bool verif = impl::CPreInsertionVerifier<ContainerT_T, typename Base::ContainerT, std::is_base_of<impl::CMultiCache_tag, MY_TYPE>::value>::verify(res);\
+    if (verif)\
+        greet(_val);\
+    return verif;
+
+    template<
+        bool isMultiContainer,
```

Comment:
this template paramter name is misleading, I think you meant `isNotVectorBased`

Reply:
I meant "is multimal or unordered_multimap". isNotVectorBased would be misleading  here because it doesn't tell that only "multi" containers are allowed here (and vector)